net.rocketpool/image/src/app.ts

```diff
@@ -1,163 +1,262 @@
+    );
+
+    if (stopProcessing) {
+      return false;
+    }
   } while (toBlock != endBlock);
+
+  return true;
 }
 
 /////////////////////////////////////////////////////////////////////////////////////////
 
 async function readEventsFromETHRangeWithFallback(
-  fromBlock,
-  toBlock,
-  clb: (_: Event) => boolean
-) {
-  try {
-    const logs = await web3.eth.getPastLogs({
-      fromBlock,
-      toBlock,
-      address: rethAddress,
-    });
-
-    const logsDecoded = abiDecoder.decodeLogs(logs);
+  rethContract: RethAbi,
+  fromBlock: number,
+  toBlock: number,
+  onEventCallback: OnEventCallback
+): Promise<AllBlockRangeHasProcessed> {
+  const eventFilters = [
+    rethContract.filters.TokensMinted(),
+    rethContract.filters.TokensBurned()
+  ] as const;
+  const eventsRowsPromises = eventFilters.map(async (eventFilter) => {
+    const events = await rethContract.queryFilter(eventFilter, fromBlock, toBlock);
 
-    for (let i = 0; i != logsDecoded.length; i++) {
-      const logDecoded = logsDecoded[i];
-      const logRaw = logs[i];
+    return events.map((event) => {
+      return {
+        eventName: event.event,
+        amount: event.args.amount.toString(),
+        ethAmount: event.args.ethAmount.toString(),
+        blockNumber: event.blockNumber,
+        blockHash: event.blockHash,
+        eventTime: event.args.time.toNumber(),
+        transactionIndex: event.transactionIndex,
+        transactionHash: event.transactionHash,
+        logIndex: event.logIndex,
+      } as EventRow;
+    });
+  });
+  const eventRows = (await Promise.all(eventsRowsPromises)).flat();
 
-      if (!["TokensMinted", "TokensBurned"].includes(logDecoded.name)) {
-        continue;
-      }
+  eventRows.sort((right, left) => right.eventTime - left.eventTime);
 
-      const event = {
-        eventName: logDecoded.name,
-        amount: logDecoded.events.find((i) => i.name == "amount").value,
-        ethAmount: logDecoded.events.find((i) => i.name == "ethAmount").value,
-        blockNumber: logRaw.blockNumber,
-        blockHash: logRaw.blockHash,
-        blockTime: Number(
-          logDecoded.events.find((i) => i.name == "time").value
-        ),
-        transactionIndex: logRaw.transactionIndex,
-        transactionHash: logRaw.transactionHash,
-        logIndex: logRaw.logIndex,
-      };
-
-      if (!clb(event)) {
-        return;
-      }
-    }
-  } catch (err) {
-    // TODO: Is this too Infura-specific?
-    if (err.message.includes("query returned more than")) {
-      const middle = Math.round((fromBlock + toBlock) / 2);
-      console.log(
-        "Hit API limit, splitting interval to [%s, %s] and [%s, %s]",
-        fromBlock,
-        middle,
-        middle + 1,
-        toBlock
-      );
-
-      await readEventsFromETHRangeWithFallback(fromBlock, middle, clb);
-      await readEventsFromETHRangeWithFallback(middle + 1, toBlock, clb);
-    } else {
-      throw err;
+  for (const eventRow of eventRows) {
```

Comment:
Optional:
you can change to to return `!!eventRows.find((eventRow) =>  !onEventCallback(eventRow));`

Reply:
Personally, I like FP approaches but in this case double negations looks a bit tricky, so I'd like propose leave as is